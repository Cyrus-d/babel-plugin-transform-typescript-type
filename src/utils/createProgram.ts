import * as ts from 'typescript';
import * as fs from 'fs';
import * as path from 'path';

const tsconfig: ts.CompilerOptions = {
  emitDecoratorMetadata: true,
  experimentalDecorators: true,
  module: ts.ModuleKind.CommonJS,
  noEmit: true,
  strictNullChecks: false,
  target: ts.ScriptTarget.ES5,
};

const fixFileName = (fileName: string) => fileName.split('/').join('\\');

const cacheProgram = ts.createProgram([''], tsconfig);

const cacheSourceFiles = cacheProgram.getSourceFiles().reduce((obj, sourceFile) => {
  obj[sourceFile.fileName] = sourceFile;

  return obj;
}, {} as SourceFileObject);

const getSourceFile = (fileName: string) => {
  if (path.isAbsolute(fileName)) {
    return cacheSourceFiles[fileName];
  }

  const file = Object.keys(cacheSourceFiles).find(x => x.endsWith(fileName));
  if (file) return cacheSourceFiles[file];

  return undefined;
};

interface SourceFileObject {
  [key: string]: ts.SourceFile;
}

const getSourceFileText = (fileName: string) => {
  if (!fs.existsSync(fileName)) return null;

  return fs.readFileSync(fileName, 'utf8');
};

const createSourceFile = (fileName: string, requestedFileName?: string, forceUpdate?: boolean) => {
  fileName = fixFileName(fileName);
  if (!forceUpdate) {
    const sourceFile = getSourceFile(fileName);
    if (sourceFile) {
      return sourceFile;
    }
  }

  const text = getSourceFileText(fileName);
  if (text === null) return undefined;
  const sourceFile = ts.createSourceFile(fileName, text, tsconfig.target!, true, ts.ScriptKind.TS);
  cacheSourceFiles[fileName] = sourceFile;

  return sourceFile;
};

export const updateSourceFileByPath = (path: string) => {
  createSourceFile(path, undefined, true);
};

export const createProgram = (filePath: string, propName: string) => {
  const program = ts.createProgram([filePath], tsconfig, {
    fileExists(fileName): boolean {
      const file = createSourceFile(fileName, filePath);
      if (!file) {
        return false;
      }

      return true;
    },
    getCanonicalFileName(_fileName: string): string {
      return _fileName;
    },
    getCurrentDirectory(): string {
      return '';
    },
    getDefaultLibFileName(): string {
      return ts.getDefaultLibFileName(tsconfig);
    },

    getNewLine(): string {
      return '\r\n';
    },
    getSourceFile(fileName): ts.SourceFile | undefined {
      return createSourceFile(fileName, filePath);
    },
    readFile(fileName: string): string {
      const src = createSourceFile(fileName, filePath);
      if (!src) return '';

      return src.text;
    },
    useCaseSensitiveFileNames(): boolean {
      return false;
    },
    writeFile: (_fileName, _content) => null,
  });

  return program;
};
