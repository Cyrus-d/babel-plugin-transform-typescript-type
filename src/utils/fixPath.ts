export const fixPath = (fileName: string) => fileName.split('/').join('\\');
