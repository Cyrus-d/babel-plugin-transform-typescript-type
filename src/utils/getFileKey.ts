export const getFileKey = (fileName: string) =>
  fileName
    .split('/')
    .join('_')
    .split('\\')
    .join('_');
