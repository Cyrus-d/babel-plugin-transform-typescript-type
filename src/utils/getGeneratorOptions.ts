import { types as t } from '@babel/core';
import { GeneratorOptions } from '../typings';

interface Option {
  [key: string]: any;
}

const setArrayOption = (prop: t.ObjectProperty, opts: Option) => {
  if (!t.isArrayExpression(prop.value)) return;

  if (t.isIdentifier(prop.key)) {
    opts[prop.key.name] = prop.value.elements.reduce((arr, elm) => {
      if (t.isStringLiteral(elm)) arr.push(elm.value);

      return arr;
    }, [] as string[]);
  }
};

const setOtherOptions = (prop: t.ObjectProperty, opts: Option) => {
  if (
    !t.isNumericLiteral(prop.value) &&
    !t.isStringLiteral(prop.value) &&
    !t.isBooleanLiteral(prop.value)
  )
    return;

  if (t.isIdentifier(prop.key)) {
    opts[prop.key.name] = prop.value.value;
  }
};

export const getGeneratorOptions = <T extends t.CallExpression>(node: T): GeneratorOptions<T> => {
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  if (!node.arguments) return {} as GeneratorOptions<T>;

  return node.arguments.reduce((opts, arg) => {
    if (t.isObjectExpression(arg)) {
      arg.properties.forEach(prop => {
        if (t.isObjectProperty(prop)) {
          setArrayOption(prop, opts);
          setOtherOptions(prop, opts);
        }
      });
    }

    return opts;
  }, {} as GeneratorOptions<{}>);
};
