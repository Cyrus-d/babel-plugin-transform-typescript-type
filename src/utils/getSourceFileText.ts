import * as fs from 'fs';

export const getSourceFileText = (fileName: string) => {
  if (!fs.existsSync(fileName)) return null;

  return fs.readFileSync(fileName, 'utf8');
};
