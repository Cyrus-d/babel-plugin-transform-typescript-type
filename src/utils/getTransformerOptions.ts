import { types as t } from '@babel/core';
import { TransformerOptions } from '../typings';

interface Option {
  [key: string]: any;
}

const setArrayOption = (prop: t.ObjectProperty, opts: Option) => {
  if (!t.isArrayExpression(prop.value)) return;

  if (t.isIdentifier(prop.key)) {
    opts[prop.key.name] = prop.value.elements.reduce((arr, elm) => {
      if (t.isStringLiteral(elm)) arr.push(elm.value);

      return arr;
    }, [] as string[]);
  }
};

const setOtherOptions = (prop: t.ObjectProperty, opts: Option) => {
  if (
    !t.isNumericLiteral(prop.value) &&
    !t.isStringLiteral(prop.value) &&
    !t.isBooleanLiteral(prop.value)
  )
    return;

  if (t.isIdentifier(prop.key)) {
    opts[prop.key.name] = prop.value.value;
  }
};

export const getTransformerOptions = <T extends t.CallExpression>(
  node: T,
): TransformerOptions<T> => {
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  if (!node.arguments) return {} as TransformerOptions<T>;

  return node.arguments.reduce((opts, arg) => {
    if (t.isObjectExpression(arg)) {
      arg.properties.forEach(prop => {
        if (t.isObjectProperty(prop)) {
          setArrayOption(prop, opts);
          setOtherOptions(prop, opts);
        }
      });
    }

    return opts;
  }, {} as TransformerOptions<{}>);
};
