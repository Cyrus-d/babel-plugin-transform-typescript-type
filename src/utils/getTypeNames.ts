import { types as t } from '@babel/core';

const getTSTypeReferenceTypeName = (node: t.TSTypeReference) => {
  if (t.isIdentifier(node.typeName)) {
    return node.typeName.name;
  }

  return null;
};

const getIntersectionTypeNames = (node: t.TSIntersectionType) => {
  const typeNames = node.types.reduce((arr, type) => {
    if (t.isTSTypeReference(type)) {
      const typeName = getTSTypeReferenceTypeName(type);
      if (typeName) arr.push(typeName);
    }

    return arr;
  }, [] as string[]);

  return typeNames;
};

export const getTsTypeName = (node: t.TSType) => {
  let typeNames: string[] = [];
  if (t.isTSTypeReference(node)) {
    const typeName = getTSTypeReferenceTypeName(node);
    if (typeName) typeNames.push(typeName);
  } else if (t.isTSIntersectionType(node)) {
    typeNames = getIntersectionTypeNames(node);
  }

  return typeNames;
};

const getTypeParametersTypeNames = (typeParameters: t.TSTypeParameterInstantiation) => {
  return typeParameters.params.reduce((arr, param) => {
    return getTsTypeName(param);
  }, [] as string[]);
};

export const getNodeTypesNames = (node: t.Node) => {
  if (t.isCallExpression(node)) {
    return getTypeParametersTypeNames((node as any).typeParameters);
  }

  return [];
};
