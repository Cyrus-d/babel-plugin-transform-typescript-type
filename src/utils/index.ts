export * from './createProgram';
export * from './fixPath';
export * from './getTransformerOptions';
export * from './getTypeNames';
export * from './mergeSchema';
export * from './moduleDependencies';
export * from './updateReferences';
