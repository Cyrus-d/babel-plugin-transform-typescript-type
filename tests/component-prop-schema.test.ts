import path from 'path';
import glob from 'fast-glob';
import { transform } from './utils';

describe('prop schema generator', () => {
  glob
    .sync('./fixtures/manual-converter/component-prop-schema/**/*.{ts,tsx}', {
      cwd: __dirname,
      dot: false,
    })
    .forEach(basePath => {
      const filePath = String(basePath);

      if (filePath.includes('/special/') || filePath.includes('/typings/')) {
        return;
      }
      it(`transforms ${filePath}`, () => {
        expect(
          transform(path.join(__dirname, filePath), {}, { transformReactPropTypesManually: true }),
        ).toMatchSnapshot();
      });
    });

  it('should generate prop schema in production', () => {
    expect(
      transform(
        path.join(
          __dirname,
          './fixtures/manual-converter/component-prop-schema/special/component-prop-schema-production.ts',
        ),
        {},
        {
          isProduction: true,
        },
      ),
    ).toMatchSnapshot();
  });

  it('should always generate prop schema in production', () => {
    expect(
      transform(
        path.join(
          __dirname,
          './fixtures/manual-converter/component-prop-schema/special/component-prop-schema-production-global.ts',
        ),
        {},
        {
          isProduction: true,
          transformReactPropTypesManually: true,
          transformReactPropSchemaInProduction: true,
        },
      ),
    ).toMatchSnapshot();
  });
});
