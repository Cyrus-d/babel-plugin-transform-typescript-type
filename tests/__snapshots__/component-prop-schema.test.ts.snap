// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prop schema generator should always generate prop schema in production 1`] = `
"import React, { Component, SFC, memo, forwardRef, Ref } from 'react';
import { transformComponentPropsToSchema } from '../../../../../src/transformerFunctions';
export interface ComponentProps {
  prop_a: string;
  prop_b?: boolean;
}

const FunctionComponent = (props: ComponentProps) => {
  return null;
};

FunctionComponent.__propsSchema = {
  'type': 'object',
  'properties': {
    'prop_a': {
      'type': 'string'
    },
    'prop_b': {
      'type': 'boolean'
    }
  },
  'required': ['prop_a'],
  'additionalProperties': false
};"
`;

exports[`prop schema generator should generate prop schema in production 1`] = `
"import React, { Component, SFC, memo, forwardRef, Ref } from 'react';
import { transformComponentPropsToSchema } from '../../../../../src/transformerFunctions';
export interface ComponentProps {
  prop_a: string;
  prop_b?: boolean;
}

const FunctionComponent = (props: ComponentProps) => {
  return null;
};

FunctionComponent.__propsSchema = {
  'type': 'object',
  'properties': {
    'prop_a': {
      'type': 'string'
    },
    'prop_b': {
      'type': 'boolean'
    }
  },
  'required': ['prop_a'],
  'additionalProperties': false
};"
`;

exports[
  `prop schema generator transforms ./fixtures/manual-converter/component-prop-schema/component-empty-interface.ts 1`
] = `
"import React, { Component, SFC, memo, forwardRef, Ref } from 'react';
import { transformComponentPropsToSchema } from '../../../../src/transformerFunctions';
export interface ComponentProps {}

const FunctionComponent = (props: ComponentProps) => {
  return null;
};

const SFCComponent: SFC<ComponentProps> = props => {
  return null;
};

class ClassComponent extends Component<ComponentProps> {
  render() {
    return null;
  }

}

const MemoComponent: SFC<ComponentProps> = memo(props => {
  return null;
});
const ForwardRefComponent: SFC<ComponentProps> = forwardRef((props, ref: Ref<HTMLDivElement>) => {
  return null;
});
const MemoForwardRefComponent: SFC<ComponentProps> = memo(forwardRef((props, ref: Ref<HTMLDivElement>) => {
  return null;
}));"
`;

exports[
  `prop schema generator transforms ./fixtures/manual-converter/component-prop-schema/component-extends-props.ts 1`
] = `
"import React, { Component, SFC, memo, Ref, forwardRef } from 'react';
import { transformComponentPropsToSchema } from '../../../../src/transformerFunctions';
import { ExternalProps } from '../typings';
export interface ComponentProps extends ExternalProps {
  prop_a: string;
  prop_b?: boolean;
}

const FunctionComponent = (props: ComponentProps) => {
  return null;
};

FunctionComponent.__propsSchema = {
  'type': 'object',
  'properties': {
    'ext_prop_1': {
      'type': 'string'
    },
    'ext_prop_2': {
      'type': 'string'
    },
    'prop_a': {
      'type': 'string'
    },
    'prop_b': {
      'type': 'boolean'
    }
  },
  'required': ['ext_prop_1', 'ext_prop_2', 'prop_a'],
  'additionalProperties': false
};

const SFCComponent: SFC<ComponentProps> = props => {
  return null;
};

SFCComponent.__propsSchema = {
  'type': 'object',
  'properties': {
    'ext_prop_1': {
      'type': 'string'
    },
    'ext_prop_2': {
      'type': 'string'
    },
    'prop_a': {
      'type': 'string'
    },
    'prop_b': {
      'type': 'boolean'
    }
  },
  'required': ['ext_prop_1', 'ext_prop_2', 'prop_a'],
  'additionalProperties': false
};

class ClassComponent extends Component<ComponentProps> {
  render() {
    return null;
  }

}

ClassComponent.__propsSchema = {
  'type': 'object',
  'properties': {
    'ext_prop_1': {
      'type': 'string'
    },
    'ext_prop_2': {
      'type': 'string'
    },
    'prop_a': {
      'type': 'string'
    },
    'prop_b': {
      'type': 'boolean'
    }
  },
  'required': ['ext_prop_1', 'ext_prop_2', 'prop_a'],
  'additionalProperties': false
};
const MemoComponent: SFC<ComponentProps> = memo(props => {
  return null;
});
MemoComponent.__propsSchema = {
  'type': 'object',
  'properties': {
    'ext_prop_1': {
      'type': 'string'
    },
    'ext_prop_2': {
      'type': 'string'
    },
    'prop_a': {
      'type': 'string'
    },
    'prop_b': {
      'type': 'boolean'
    }
  },
  'required': ['ext_prop_1', 'ext_prop_2', 'prop_a'],
  'additionalProperties': false
};
const ForwardRefComponent: SFC<ComponentProps> = forwardRef((props, ref: Ref<HTMLDivElement>) => {
  return null;
});
ForwardRefComponent.__propsSchema = {
  'type': 'object',
  'properties': {
    'ext_prop_1': {
      'type': 'string'
    },
    'ext_prop_2': {
      'type': 'string'
    },
    'prop_a': {
      'type': 'string'
    },
    'prop_b': {
      'type': 'boolean'
    }
  },
  'required': ['ext_prop_1', 'ext_prop_2', 'prop_a'],
  'additionalProperties': false
};
const MemoForwardRefComponent: SFC<ComponentProps> = memo(forwardRef((props, ref: Ref<HTMLDivElement>) => {
  return null;
}));
MemoForwardRefComponent.__propsSchema = {
  'type': 'object',
  'properties': {
    'ext_prop_1': {
      'type': 'string'
    },
    'ext_prop_2': {
      'type': 'string'
    },
    'prop_a': {
      'type': 'string'
    },
    'prop_b': {
      'type': 'boolean'
    }
  },
  'required': ['ext_prop_1', 'ext_prop_2', 'prop_a'],
  'additionalProperties': false
};"
`;

exports[
  `prop schema generator transforms ./fixtures/manual-converter/component-prop-schema/component-mix-props.ts 1`
] = `
"import React, { Component, SFC, memo, forwardRef, Ref } from 'react';
import { transformComponentPropsToSchema } from '../../../../src/transformerFunctions';
import { ExternalProps } from '../typings';
export interface ComponentProps {
  prop_a: string;
  prop_b?: boolean;
}

const FunctionComponent = (props: ComponentProps & ExternalProps) => {
  return null;
};

FunctionComponent.__propsSchema = {
  'type': 'object',
  'properties': {
    'prop_a': {
      'type': 'string'
    },
    'prop_b': {
      'type': 'boolean'
    },
    'ext_prop_1': {
      'type': 'string'
    },
    'ext_prop_2': {
      'type': 'string'
    }
  },
  'required': ['prop_a', 'ext_prop_1', 'ext_prop_2'],
  'additionalProperties': false
};

const SFCComponent: SFC<ComponentProps & ExternalProps> = props => {
  return null;
};

SFCComponent.__propsSchema = {
  'type': 'object',
  'properties': {
    'prop_a': {
      'type': 'string'
    },
    'prop_b': {
      'type': 'boolean'
    },
    'ext_prop_1': {
      'type': 'string'
    },
    'ext_prop_2': {
      'type': 'string'
    }
  },
  'required': ['prop_a', 'ext_prop_1', 'ext_prop_2'],
  'additionalProperties': false
};

class ClassComponent extends Component<ComponentProps & ExternalProps> {
  render() {
    return null;
  }

}

ClassComponent.__propsSchema = {
  'type': 'object',
  'properties': {
    'prop_a': {
      'type': 'string'
    },
    'prop_b': {
      'type': 'boolean'
    },
    'ext_prop_1': {
      'type': 'string'
    },
    'ext_prop_2': {
      'type': 'string'
    }
  },
  'required': ['prop_a', 'ext_prop_1', 'ext_prop_2'],
  'additionalProperties': false
};
const MemoComponent: SFC<ComponentProps & ExternalProps> = memo(props => {
  return null;
});
MemoComponent.__propsSchema = {
  'type': 'object',
  'properties': {
    'prop_a': {
      'type': 'string'
    },
    'prop_b': {
      'type': 'boolean'
    },
    'ext_prop_1': {
      'type': 'string'
    },
    'ext_prop_2': {
      'type': 'string'
    }
  },
  'required': ['prop_a', 'ext_prop_1', 'ext_prop_2'],
  'additionalProperties': false
};
const ForwardRefComponent: SFC<ComponentProps & ExternalProps> = forwardRef((props, ref: Ref<HTMLDivElement>) => {
  return null;
});
ForwardRefComponent.__propsSchema = {
  'type': 'object',
  'properties': {
    'prop_a': {
      'type': 'string'
    },
    'prop_b': {
      'type': 'boolean'
    },
    'ext_prop_1': {
      'type': 'string'
    },
    'ext_prop_2': {
      'type': 'string'
    }
  },
  'required': ['prop_a', 'ext_prop_1', 'ext_prop_2'],
  'additionalProperties': false
};
const MemoForwardRefComponent: SFC<ComponentProps & ExternalProps> = memo(forwardRef((props, ref: Ref<HTMLDivElement>) => {
  return null;
}));
MemoForwardRefComponent.__propsSchema = {
  'type': 'object',
  'properties': {
    'prop_a': {
      'type': 'string'
    },
    'prop_b': {
      'type': 'boolean'
    },
    'ext_prop_1': {
      'type': 'string'
    },
    'ext_prop_2': {
      'type': 'string'
    }
  },
  'required': ['prop_a', 'ext_prop_1', 'ext_prop_2'],
  'additionalProperties': false
};"
`;

exports[
  `prop schema generator transforms ./fixtures/manual-converter/component-prop-schema/component-no-props.ts 1`
] = `
"import React, { Component, SFC, memo, forwardRef, Ref } from 'react';
import { transformComponentPropsToSchema } from '../../../../src/transformerFunctions';

const FunctionComponent = props => {
  return null;
};

const SFCComponent = props => {
  return null;
};

class ClassComponent extends Component {
  render() {
    return null;
  }

}

const MemoComponent = memo(props => {
  return null;
});
const ForwardRefComponent = forwardRef((props, ref: Ref<HTMLDivElement>) => {
  return null;
});
const MemoForwardRefComponent = memo(forwardRef((props, ref: Ref<HTMLDivElement>) => {
  return null;
}));"
`;

exports[
  `prop schema generator transforms ./fixtures/manual-converter/component-prop-schema/component-prop-schema.ts 1`
] = `
"import React, { Component, SFC, memo, forwardRef, Ref } from 'react';
import { transformComponentPropsToSchema } from '../../../../src/transformerFunctions';
export interface ComponentProps {
  prop_a: string;
  prop_b?: boolean;
}

const FunctionComponent = (props: ComponentProps) => {
  return null;
};

FunctionComponent.__propsSchema = {
  'type': 'object',
  'properties': {
    'prop_a': {
      'type': 'string'
    },
    'prop_b': {
      'type': 'boolean'
    }
  },
  'required': ['prop_a'],
  'additionalProperties': false
};

const SFCComponent: SFC<ComponentProps> = props => {
  return null;
};

SFCComponent.__propsSchema = {
  'type': 'object',
  'properties': {
    'prop_a': {
      'type': 'string'
    },
    'prop_b': {
      'type': 'boolean'
    }
  },
  'required': ['prop_a'],
  'additionalProperties': false
};

class ClassComponent extends Component<ComponentProps> {
  render() {
    return null;
  }

}

ClassComponent.__propsSchema = {
  'type': 'object',
  'properties': {
    'prop_a': {
      'type': 'string'
    },
    'prop_b': {
      'type': 'boolean'
    }
  },
  'required': ['prop_a'],
  'additionalProperties': false
};
const MemoComponent: SFC<ComponentProps> = memo(props => {
  return null;
});
MemoComponent.__propsSchema = {
  'type': 'object',
  'properties': {
    'prop_a': {
      'type': 'string'
    },
    'prop_b': {
      'type': 'boolean'
    }
  },
  'required': ['prop_a'],
  'additionalProperties': false
};
const ForwardRefComponent: SFC<ComponentProps> = forwardRef((props, ref: Ref<HTMLDivElement>) => {
  return null;
});
ForwardRefComponent.__propsSchema = {
  'type': 'object',
  'properties': {
    'prop_a': {
      'type': 'string'
    },
    'prop_b': {
      'type': 'boolean'
    }
  },
  'required': ['prop_a'],
  'additionalProperties': false
};
const MemoForwardRefComponent: SFC<ComponentProps> = memo(forwardRef((props, ref: Ref<HTMLDivElement>) => {
  return null;
}));
MemoForwardRefComponent.__propsSchema = {
  'type': 'object',
  'properties': {
    'prop_a': {
      'type': 'string'
    },
    'prop_b': {
      'type': 'boolean'
    }
  },
  'required': ['prop_a'],
  'additionalProperties': false
};"
`;

exports[
  `prop schema generator transforms ./fixtures/manual-converter/component-prop-schema/component-type-schema.ts 1`
] = `
"import React, { Component, SFC, memo, forwardRef, Ref } from 'react';
import { transformComponentPropsToSchema } from '../../../../src/transformerFunctions';
import { ExternalProps } from '../typings';
export interface ComponentProps {
  prop_a: string;
  prop_b?: boolean;
}
type OwnProps = ComponentProps & ExternalProps;

const FunctionComponent = (props: OwnProps) => {
  return null;
};

FunctionComponent.__propsSchema = {
  'type': 'object',
  'additionalProperties': false,
  'properties': {
    'ext_prop_1': {
      'type': 'string'
    },
    'ext_prop_2': {
      'type': 'string'
    },
    'prop_a': {
      'type': 'string'
    },
    'prop_b': {
      'type': 'boolean'
    }
  },
  'required': ['ext_prop_1', 'ext_prop_2', 'prop_a']
};

const SFCComponent: SFC<OwnProps> = props => {
  return null;
};

SFCComponent.__propsSchema = {
  'type': 'object',
  'additionalProperties': false,
  'properties': {
    'ext_prop_1': {
      'type': 'string'
    },
    'ext_prop_2': {
      'type': 'string'
    },
    'prop_a': {
      'type': 'string'
    },
    'prop_b': {
      'type': 'boolean'
    }
  },
  'required': ['ext_prop_1', 'ext_prop_2', 'prop_a']
};

class ClassComponent extends Component<OwnProps> {
  render() {
    return null;
  }

}

ClassComponent.__propsSchema = {
  'type': 'object',
  'additionalProperties': false,
  'properties': {
    'ext_prop_1': {
      'type': 'string'
    },
    'ext_prop_2': {
      'type': 'string'
    },
    'prop_a': {
      'type': 'string'
    },
    'prop_b': {
      'type': 'boolean'
    }
  },
  'required': ['ext_prop_1', 'ext_prop_2', 'prop_a']
};
const MemoComponent: SFC<OwnProps> = memo(props => {
  return null;
});
MemoComponent.__propsSchema = {
  'type': 'object',
  'additionalProperties': false,
  'properties': {
    'ext_prop_1': {
      'type': 'string'
    },
    'ext_prop_2': {
      'type': 'string'
    },
    'prop_a': {
      'type': 'string'
    },
    'prop_b': {
      'type': 'boolean'
    }
  },
  'required': ['ext_prop_1', 'ext_prop_2', 'prop_a']
};
const ForwardRefComponent: SFC<OwnProps> = forwardRef((props, ref: Ref<HTMLDivElement>) => {
  return null;
});
ForwardRefComponent.__propsSchema = {
  'type': 'object',
  'additionalProperties': false,
  'properties': {
    'ext_prop_1': {
      'type': 'string'
    },
    'ext_prop_2': {
      'type': 'string'
    },
    'prop_a': {
      'type': 'string'
    },
    'prop_b': {
      'type': 'boolean'
    }
  },
  'required': ['ext_prop_1', 'ext_prop_2', 'prop_a']
};
const MemoForwardRefComponent: SFC<OwnProps> = memo(forwardRef((props, ref: Ref<HTMLDivElement>) => {
  return null;
}));
MemoForwardRefComponent.__propsSchema = {
  'type': 'object',
  'additionalProperties': false,
  'properties': {
    'ext_prop_1': {
      'type': 'string'
    },
    'ext_prop_2': {
      'type': 'string'
    },
    'prop_a': {
      'type': 'string'
    },
    'prop_b': {
      'type': 'boolean'
    }
  },
  'required': ['ext_prop_1', 'ext_prop_2', 'prop_a']
};"
`;

exports[
  `prop schema generator transforms ./fixtures/manual-converter/component-prop-schema/excludeProps-options.ts 1`
] = `
"import React, { Component, SFC, memo, Ref, forwardRef } from 'react';
import { transformComponentPropsToSchema } from '../../../../src/transformerFunctions';
export interface ComponentProps {
  prop_a: string;
  prop_b: string;
}

const FunctionComponent = (props: ComponentProps) => {
  return null;
};

FunctionComponent.__propsSchema = {
  'type': 'object',
  'properties': {
    'prop_a': {
      'type': 'string'
    }
  },
  'required': ['prop_a'],
  'additionalProperties': false
};"
`;

exports[
  `prop schema generator transforms ./fixtures/manual-converter/component-prop-schema/includeProps-options.ts 1`
] = `
"import React, { Component, SFC, memo, Ref, forwardRef } from 'react';
import { transformComponentPropsToSchema } from '../../../../src/transformerFunctions';
export interface ComponentProps {
  prop_a: string;
  prop_b: string;
}

const FunctionComponent = (props: ComponentProps) => {
  return null;
};

FunctionComponent.__propsSchema = {
  'type': 'object',
  'properties': {
    'prop_b': {
      'type': 'string'
    }
  },
  'required': ['prop_b'],
  'additionalProperties': false
};"
`;

exports[
  `prop schema generator transforms ./fixtures/manual-converter/component-prop-schema/should-applay-maxDept.ts 1`
] = `
"import React, { Component, SFC, memo, Ref, forwardRef } from 'react';
import { transformComponentPropsToSchema } from '../../../../src/transformerFunctions';
interface InnerProps {
  level_1: {
    level_2: {
      level_3: string;
    };
  };
}
export interface ComponentProps {
  level_0: InnerProps;
}

const FunctionComponent = (props: ComponentProps) => {
  return null;
};

FunctionComponent.__propsSchema = {
  'type': 'object',
  'properties': {
    'level_0': {
      'type': 'object',
      'properties': {
        'level_1': {
          'type': 'object',
          'additionalProperties': false
        }
      },
      'required': ['level_1'],
      'additionalProperties': false
    }
  },
  'required': ['level_0'],
  'additionalProperties': false
};"
`;
